### PR TITLE
guides: add new generator

### PR DESCRIPTION
**Pull request by millotp**
## 🧭 What and Why

🎟 JIRA Ticket: https://algolia.atlassian.net/browse/DI-2974 https://algolia.atlassian.net/browse/DI-2975 https://algolia.atlassian.net/browse/DI-2976

### Changes included:

supersede https://github.com/algolia/api-clients-automation/pull/3836

this PR adds a new generator in order to be able to generate guides.

### Context

The onboarding team creates guides using the API client plus some other external libraries (e.g. fetch) in order to showcase how to push data to Algolia.

Today they only provide snippets for JavaScript, and it requires us manual intervention, manual testing, linting etc.

They'd like to be able to keep their guides up to dates with our API clients, and at some point, provide guides for more languages than just JavaScript.

In their initial RFC, they were proposing to create a CI that does exactly what we do, so in order to platformize a bit more our repository, @shortcuts suggested to leverage our tooling to generate their guides

### How does it work

This PR is not complete, we are missing a few things that will be tackled in https://algolia.atlassian.net/browse/DI-2984 and https://algolia.atlassian.net/browse/DI-2985 but the goal is simple

From a mustache file (located in `templates/${language}/guides/${api}`), and a guide definition (located in `tests/CTS/guides/${api}`), generate a fully working guide, self contained file, for every available clients.

### Example

I create my `templates/go/guides/search/saveObjectsWithSofa.mustache` file:

```mustache
{{> snippets/import}}

{{> snippets/init}}

// do boilerplate stuff with sofa
return {{#dynamicSnippet}}saveObjectsWithSofa{{/dynamicSnippet}}
```

The dynamic snippet template will read the `tests/CTS/guides/search/snippets.json`:

```json
{
  "saveObjectsWithSofa": {
    "method": "saveObjects",
    "parameters": {
      "indexName": "sofaaa",
      "objects": [
        {
          "$var": "objects"
        }
      ]
    }
  }
}
```

The key that has been passed to the dynamic template (`saveObjectsWithSofa`) is there and contains the method to find in the spec, and the parameters to feed it. The `parametersWithDataType` logic is the same as for the tests and snippets generator, as well as for the mustaches file.

You can now write for example `templates/php/guides/search/saveObjectsWithSofa.mustache`, the process will be the same.

### Other changes

We don't need to have the snippets folder/files anymore, just the JSON file output for the documentation. So we removed it from the repo, and ignored it, it now only lives on the CI

### CODE REVIEW

File: tests/CTS/guides/search/snippets.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+          "objectID": "1",
+          "name": "Adam"
+        },
+        {
+          "objectID": "2",
+          "name": "Benoit"
+        }
+      ]
+    }
+  }
```
Comment (shortcuts) [inline tests/CTS/guides/search/snippets.json:33]: yes

File: .gitignore [hunk 2]
```diff
@@ -7,6 +7,7 @@ node_modules
 yarn-error.log
 debug.log
 
+
```
Comment (shortcuts) [inline .gitignore:10]:
```suggestion
```

File: tests/CTS/guides/search/snippets.json
Comment (shortcuts) [inline tests/CTS/guides/search/snippets.json:1]: added a ticket to document the schema https://algolia.atlassian.net/browse/DI-2985

File: templates/javascript/snippets/package.mustache [hunk 3]
```diff
@@ -1,7 +1,8 @@
 {
-  "name": "javascript-snippets",
+  "name": "javascript-{{mode}}",
   "version": "1.0.0",
   "dependencies": {
+    "typescript": "5.6.2",
```
Comment (shortcuts) [inline templates/javascript/snippets/package.mustache:5]: do we need to setup renovate?

Comment (millotp) [reply templates/javascript/snippets/package.mustache:5]: it's already setup !

**Comment by shortcuts** [comment]
of course it's perfect

File: generators/src/main/java/com/algolia/codegen/cts/manager/SwiftCTSManager.java [hunk 4]
```diff
@@ -41,7 +41,9 @@ public void addDataToBundle(Map<String, Object> bundle) throws GeneratorExceptio
   }
 
   @Override
-  public void addSnippetsSupportingFiles(List<SupportingFile> supportingFiles) {
-    supportingFiles.add(new SupportingFile("snippets/Package.mustache", "snippets/swift", "Package.swift"));
+  public void addSnippetsSupportingFiles(List<SupportingFile> supportingFiles, String output) {
+    supportingFiles.add(new SupportingFile("snippets/Package.mustache", output + "/swift", "Package.swift"));
+    supportingFiles.add(new SupportingFile("snippets/.swiftformat.mustache", output + "/swift", ".siwftformat"));
```
Comment (Fluf22) [inline generators/src/main/java/com/algolia/codegen/cts/manager/SwiftCTSManager.java:46]:
```suggestion
    supportingFiles.add(new SupportingFile("snippets/.swiftformat.mustache", output + "/swift", ".swiftformat"));
```

Comment (shortcuts) [reply generators/src/main/java/com/algolia/codegen/cts/manager/SwiftCTSManager.java:46]: oh wow how come it works ahah

Comment (Fluf22) [reply generators/src/main/java/com/algolia/codegen/cts/manager/SwiftCTSManager.java:46]: It has default for when it doesn't detect a file, so I guess that's what happens here

**Review by shortcuts — APPROVED** [review]
This is wonderful

**Pull request merged**